src/storage/src/source/mysql/replication.rs

```diff
@@ -376,7 +374,34 @@ pub(crate) fn render<G: Scope<Timestamp = GtidPartition>>(
                             .as_ref()
                             .expect("gtid cap should be set by previous GtidEvent");
 
-                        events::handle_query_event(event, &mut repl_context, new_gtid).await?;
+                        let should_advance =
+                            events::handle_query_event(event, &mut repl_context, new_gtid).await?;
+
+                        if should_advance {
+                            if let Some(mut new_gtid) = next_gtid.take() {
+                                // Increment the transaction-id to the next GTID we should see from this source-id
+                                match new_gtid.timestamp_mut() {
+                                    GtidState::Active(time) => {
+                                        *time = time.checked_add(1).unwrap();
```

Comment:
This feels a bit weird since we are guessing the current GTID. We know that each DDL statement is assigned a GTID but how do we know we're not under counting? Is it worth doing the same experiment Nikhil did that led to the discovery of the XIDEvent with a DDL statement? There might be some other event that we are not considering that comes after a DDL statatement. Is this maybe what the END STATEMENT flag is useful for?

Reply:
The end-statement flag is just on RowsEvents -- so not available here.

We aren't guessing the current GTID since we know the GTID from the last GTIDEvent we received. In this case we are only setting `should_advance` for query events that we know will trigger a new GTID, and if we under-count I don't see that as a correctness error since we would advance once we receive the subsequent GTID event.

> Is it worth doing the same experiment Nikhil did that led to the discovery of the XIDEvent with a DDL statement?

The first part of Nikhil's output https://github.com/MaterializeInc/materialize/issues/24630#issuecomment-2018082249 are two DDL statements (`create table`) and you can see there are no other binlog flags or events that occur besides the subsequent GTID event.

Reply:
Also not sure if it's obvious, but this "increment by 1" logic is the same that's done when we receive the Xid event. We are just setting the frontier beyond the current GTID since we want to expose the data/ddl changes that just occurred

Reply:
I see, ok. Undercounting is definitely an issue if it happens. It means that some event that should have timestamp `t_correct` got assigned timestmap `t_wrong < t_correct`. So if the ingestion pipeline emits `(event, t_wrong, 1)`, and then we restart at `t_wrong+1` mysql will happily resend the event to us because `t_wrong + 1 <= t_correct` and we'll emit it twice.

I don't have anything better to propose though, I was hoping there would be some additional event but indeed it doesn't appear to be the case. Thanks for the explanation!